packages/core/src/node/core-process-manager.ts

```diff
@@ -0,0 +1,97 @@
+/********************************************************************************
+ * Copyright (C) 2021 Ericsson and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import * as cp from 'child_process';
+import { injectable } from 'inversify';
+
+@injectable()
+export class CoreProcessManager {
+
+    /**
+     * Synchronously send SIGTERM to a process and its children, recursively.
+     *
+     * @param ppid The process to kill along with its children (recursive).
+     */
+    terminateProcessTree(ppid: number): void {
+        if (process.platform === 'win32') {
+            this.win_taskkillTree(ppid);
+        } else {
+            for (const pid of this.unix_getChildrenRecursive(ppid)) {
+                // Prevent killing the current process:
+                if (pid !== process.pid) {
+                    // Don't stop if a process fails to be killed (keep on killing the others):
+                    try {
+                        process.kill(pid);
+                    } catch (error) {
+                        console.error(error);
+                    }
+                }
+            }
+            if (ppid === this.unix_getpgid(ppid)) {
+                // When a process pgid === pid this means the the process is a group leader.
+                // We can then kill every process part of its group by doing `kill(-pgid)`.
+                // This can catch leaked processes under `init` that are still part of the group.
+                process.kill(-ppid);
+            }
+            process.kill(ppid);
+        }
+    }
+
+    protected unix_getpgid(pid: number): number {
+        const { stdout } = this.spawnSync('ps', ['-p', pid.toString(10), '-o', 'pgid=']);
+        return Number.parseInt(stdout, 10);
+    }
+
+    protected unix_getChildrenRecursive(ppid: number): Set<number> {
+        const { stdout } = this.spawnSync('ps', ['ax', '-o', 'ppid=,pid=']);
+        const pids = new Set<number>([ppid]);
+        const matcher = /(\d+)\s+(\d+)/;
+        const sorted = stdout
+            .trim()
+            .split('\n')
+            .map(line => line
+                .match(matcher)!.slice(1)
+                .map(value => Number.parseInt(value, 10))
+            )
+            .sort((a, b) => a[0] - b[0]);
```

Comment:
are children guaranteed to have a higher PID than their parents in general? it looks like assuming that they will lets us do this operation without recursion, which is nice, but I think there may be cases in which children have lower PID's than their parents, so we won't have seen or added the parent's entry when we see the child's line, and we therefore won't add the child or any of its children.

Reply:
Good catch, I just noticed this as well. I am working on an update.